### PR TITLE
feat: `eth_accounts` return all permitted accounts

### PR DESCRIPTION
**Pull request by jiexi**
**Development & PR Process**
1. Follow MetaMask [Mobile Coding Standards](https://docs.google.com/document/d/1VJLwTRsUw_5EDq_o8d6sSbXUAYENLurkRitYO45eY5o/edit?usp=sharing)
2. Add `release-xx` label to identify the PR slated for a upcoming release (will be used in release discussion)
3. Add `needs-dev-review` label when work is completed
4. Add `needs-qa` label when dev review is completed
5. Add `QA Passed` label when QA has signed off

**Description**

* Return all permitted accounts from `eth_accounts` with most recently used first for browser and SDK

This change was made to align with the [change suggested in extension](https://github.com/MetaMask/metamask-extension/pull/18516). The goal is start returning all accounts on each `eth_accounts` call rather than only returning an array with a single account. We achieve this by updating the `restrictReturnedAccounts` decorator

**Screenshots/Recordings**

https://github.com/MetaMask/metamask-mobile/assets/918701/215d5454-5bd1-4c07-ab04-32dc19e885ba

NOTE: test dapp page changes are not deployed yet, but have been merged into the main branch and be run locally to reproduce the recording above.

https://github.com/MetaMask/metamask-mobile/assets/918701/ff031336-b765-400d-bc7d-12d44b778315

**Issue**

[Ticket: mme-18515](https://github.com/MetaMask/metamask-extension/issues/18515)
[Documentation PR](https://github.com/MetaMask/metamask-docs/pull/773)

**Checklist**

* [x] There is a related GitHub issue
* [x] Tests are included if applicable
* [x] Any added code is fully documented

### CODE REVIEW

**Comment by jiexi** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by jiexi** [comment]
I'm unsure if my assumptions are correct for `eth_accounts` behavior: 
* Wallet Connect does not support multiple permitted accounts, and thus can only return the selectedAddress
* SDK does support multiple permitted accounts, and thus behaves the same as browser

**Comment by jiexi** [comment]
Is it a correct assumption that the mobile SDK supports multiple accounts and that mobile Wallet Connect supports only a single account?

**Comment by jiexi** [comment]
Reverted sdk and wc behavior based on discussion with @andreahaku

File: app/core/Permissions/specifications.js [hunk 1]
```diff
@@ -54,11 +54,17 @@ export const getCaveatSpecifications = ({ getIdentities }) => ({
...
-        const addressToCompare = address.toLowerCase();
-        return allAccounts.includes(addressToCompare);
-      });
-      return res.slice(0, 1);
+      const res = caveat.value
+        .filter(({ address }) => {
+          const addressToCompare = address.toLowerCase();
+          return allAccounts.includes(addressToCompare);
+        })
+        .sort((a, b) =>
```
[the remark this answers is not in the document]
Comment (Cal-L) [reply app/core/Permissions/specifications.js:62]: All you'll need to do is remove `slice`

**Review by Cal-L — COMMENTED** [review]
left a comment

**Review by Cal-L — APPROVED** [review]
LGTM

**Comment by cortisiko** [comment]
hey @jiexi I have a couple of questions:

Did you test removing accounts that were connected to a dapp?
Did you test importing accounts while one of your previous accounts was connected to a dapp?
Did you test connecting and disconnecting hardware wallets while connected to a dapp?
Was this tested against the portfolio dapp? 
WC v2 was merged a couple of days ago, was this tested on the most recent main commit?

**Comment by jiexi** [comment]
@cortisiko 

Did you test removing accounts that were connected to a dapp?
https://github.com/MetaMask/metamask-mobile/assets/918701/8a9add6f-8f57-46e0-b5af-be6b855d3893


Did you test importing accounts while one of your previous accounts was connected to a dapp?
https://github.com/MetaMask/metamask-mobile/assets/918701/ed4b203c-08c8-4769-bccb-7d9b2e808e43

Did you test connecting and disconnecting hardware wallets while connected to a dapp?
No. Don't have access to one I can use currently

Was this tested against the portfolio dapp?
Yes, I'm not sure what the correct behavior is. Seems like revoking account access doesn't remove them from the tracked accounts. But adding multiple accounts does get all the accounts to show up in the dapp on initial connect. Adding multiple accounts after the fact only gets one account added. To get the other account to show up, you have to switch and make them the active dapp. I believe this is a problem with the portfolio dapp itself specifically pulling only the active account when permissions change.

https://github.com/MetaMask/metamask-mobile/assets/918701/84e7bd23-4c6f-4405-9ddf-0016e037f27d


WC v2 was merged a couple of days ago, was this tested on the most recent main commit? 
Yes. Just getting the error from not having WALLET_CONNECT_PROJECT_ID set

**Review by cortisiko — APPROVED** [review]
Approving! I went through some manual verifications and I did not find any issues.

**Pull request merged**